map/src/main/kotlin/edu/artic/map/MapViewModel.kt

```diff
@@ -116,72 +149,86 @@ class MapViewModel @Inject constructor(
-                }
+        val objects = observeObjectsWithin(galleries)
 
         Observables.combineLatest(
-                distinctFloor,
-                zoomLevel,
                 galleries,
                 objects
-        ) { floor, zoomLevel, galleryList, objectList ->
-            return@combineLatest if (zoomLevel == MapZoomLevel.Three && floor == currentFloor) {
-                val list = mutableListOf<MapItem<*>>()
-                list.addAll(
-                        galleryList.map { gallery ->
-                            MapItem.Gallery(gallery, floor)
-                        }
+        ) { galleryList, objectList ->
+            mutableListOf<MapItem<*>>().apply {
+                addAll(
+                    galleryList.map { gallery ->
+                        MapItem.Gallery(gallery, gallery.floorAsInt)
+                    }
                 )
-                list.addAll(
-                        objectList.map { articObject ->
-                            MapItem.Object(articObject, floor)
-                        }
+                addAll(
+                        objectList
                 )
-                list
-            } else {
-                emptyList<MapItem<*>>()
             }
-        }.filter { it.isNotEmpty() }
-                .bindTo(veryDynamicMapItems)
+        }
+                // We explicitly permit emission of empty lists, as that is a signal to clear the map.
+                .bindTo(whatToDisplayOnMap)
                 .disposedBy(disposeBag)
     }
 
+    /**
+     * Observe for the Galleries which we're interested in at the moment.
+     *
+     * The returned [Observable] only emits events if the [current zoom level][zoomLevel]
+     * is set to [MapZoomLevel.Three], as that is a pre-requisite for galleries to be
+     * displayed on the map.
+     */
+    fun observeGalleriesAtFloor(): Observable<List<ArticGallery>> {
+        return observeFloorsAtZoom(MapZoomLevel.Three)
+                .flatMap {
+                    galleryDao.getGalleriesForFloor(it.toString()).toObservable()
+                }.share()
+        // Note: if we don't share this, only one observer could listen to it (we want 2 to do that)
+    }
+
+    /**
+     * Observe the [ArticObject]s in the given list of galleries.
+     *
+     * One gallery may contain multiple objects; the returned observable
+     * returns all of the objects it finds as a single list. The mechanism
+     * we use to retrieve these objects precludes the possibility of
+     * duplicates.
+     *
+     * Each of the observed [MapItem.Object]s include info about the floor
+     * of the gallery where it is found.
+     */
+    fun observeObjectsWithin(observed: Observable<List<ArticGallery>>): Observable<List<MapItem.Object>> {
+        return observed.map { galleries ->
+            galleries.filter { gallery ->
+                gallery.titleT != null
+            }.map { gallery ->
+                val title = gallery.titleT.orEmpty()
```

Comment:
switch to `title` not `titleT` for its title

Reply:
Hmm. I _was_ just preserving the previous behavior, but it doesn't look like we use `titleT` anywhere but the two places you flagged up. Let me confirm that there wouldn't be any adverse effects.

Reply:
Right, I've made the change. The new commit (174dcfca339c741c10) should resolve that.